lib/internal/modules/esm/resolve.js

```diff
@@ -195,44 +226,22 @@ function fileExists(url) {
  * @returns {URL}
  */
 function legacyMainResolve(packageJSONUrl, packageConfig, base) {
-  let guess;
-  if (packageConfig.main !== undefined) {
-    // Note: fs check redundances will be handled by Descriptor cache here.
-    if (fileExists(guess = new URL(`./${packageConfig.main}`,
-                                   packageJSONUrl))) {
-      return guess;
-    } else if (fileExists(guess = new URL(`./${packageConfig.main}.js`,
-                                          packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.node`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.js`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.node`,
-                                        packageJSONUrl)));
-    else guess = undefined;
-    if (guess) {
-      emitLegacyIndexDeprecation(guess, packageJSONUrl, base,
-                                 packageConfig.main);
-      return guess;
-    }
-    // Fallthrough.
-  }
-  if (fileExists(guess = new URL('./index.js', packageJSONUrl)));
-  // So fs.
-  else if (fileExists(guess = new URL('./index.json', packageJSONUrl)));
-  else if (fileExists(guess = new URL('./index.node', packageJSONUrl)));
-  else guess = undefined;
-  if (guess) {
-    emitLegacyIndexDeprecation(guess, packageJSONUrl, base, packageConfig.main);
-    return guess;
+  const packageJsonUrlString = packageJSONUrl.href;
+
+  if (typeof packageJsonUrlString !== 'string') {
```

Comment:
Are you sure this is correct? href attribute of a URL always returns string.

Reply:
Yes, it always returns a string, but since I didn't know for sure if packageJSONUrl is a URL, in the old code we had validations but in this one I needed to make this validation explicit.